Basestation_Software.Web/Core/Components/PingTool.razor

```diff
@@ -0,0 +1,339 @@
+
+    // Sets all components to be pinged automatically.
+    private void AutoPingAll(bool value)
+    {
+        foreach (Device device in devices)
+        {
+            device.AutoPing = value;
+        }
+    }
+
+    // Updates whether ping tool should warn user of a poor connection
+    private void UpdateWarnOfPoorConnection(bool value)
+    {
+        warnOfPoorConnection = value;
+    }
+
+    private void UpdateAutoPingDelay(string delay)
+    {
+        try
+        {
+            autoPingDelay = int.Parse(delay) * 1000;
+        }
+        catch
+        {
+            autoPingDelay = 5000;
+        }
+    }
+
+    private async void RovePingBoard(Device device, string board)
+    {
+        device.PingStatus = "Pinging...";
+        await InvokeAsync(StateHasChanged);
+
+        DateTime start = DateTime.Now;
+        RoveCommService.Ping(board);
+        var response = await RoveCommService.Listen<int>(RoveCommManifest.SystemPackets.PING_REPLY);
+        DateTime end = DateTime.Now;
+        TimeSpan latency = end - start;
+
+        if (response != null)
+        {
+            device.PingStatus = $"Response in {latency.TotalSeconds.ToString()} seconds";
+        } else
+        {
+            device.PingStatus = "NoResponse";
+        }
+        await InvokeAsync(StateHasChanged);
+    }
+
+    public class Device : INotifyPropertyChanged
+    {
+        private string pingStatus;
+        private string pingStatusHighlight;
+        private bool autoPing;
+
+        public string Name { get; set; }
+        public string IpAddress { get; set; }
+
+        public string PingStatus
+        {
+            get => pingStatus;
+            set
+            {
+                pingStatus = value;
+                OnPropertyChanged(nameof(PingStatus));
+            }
+        }
+
+        public string PingStatusHighlight
+        {
+            get => pingStatusHighlight;
+            set
+            {
+                pingStatusHighlight = value;
+                OnPropertyChanged(nameof(PingStatusHighlight));
+            }
+        }
+
+        public bool AutoPing
+        {
+            get => autoPing;
+            set
+            {
+                autoPing = value;
+                OnPropertyChanged(nameof(AutoPing));
+            }
+        }
+
+        public bool RoveCapable { get; set; }
+
+        public Device(string name, string ipAddress, bool autoPing = false, string pingStatus = "-1", bool roveCapable = false)
+        {
+            Name = name;
+            IpAddress = ipAddress;
+            AutoPing = autoPing;
+            PingStatus = pingStatus;
+            RoveCapable = roveCapable;
+            PingStatusHighlight = "background-color:lightcoral";
+        }
+
+        public event PropertyChangedEventHandler PropertyChanged;
+
+        protected void OnPropertyChanged(string propertyName)
+        {
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+        }
+    }
```

Comment:
This class should be in its own file in the .Models project and PingTool.razor should import it from there.

Reply:
The only similar classes I could find were `RoveCommBoardDesc` and `RoveCommDeviceDesc` in `RoveCommManifest.cs`. Neither of these provide the functionality I need for the specific purposes of the Ping Tool (namely, INotifyPropertyChanged and storing component-specific information such as the status highlight or auto-pinging). As such, I figured it would be best to track these component-specific variables and methods in their own class made specifically for the ping tool.

Reply:
I see. My main concern was that the Ping Graph, when it gets implemented, will likely need much of the same information and would benefit from using the above class. I think keeping in the PingTool file is fine for now, but might be changed in the future during the development of the Ping Graph.

Reply:
Sounds good.